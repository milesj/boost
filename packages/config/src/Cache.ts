import fs from 'fs';
import { Path } from '@boost/common';

export interface FileCache<T> {
	content: T;
	exists: boolean;
	mtime: number;
}

export default class Cache {
	configDir?: Path;

	dirFilesCache: Record<string, Path[]> = {};

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	fileContentCache: Record<string, FileCache<any>> = {};

	pkgPath?: Path;

	rootDir?: Path;

	async cacheFileContents<T>(path: Path, cb: () => Promise<T>): Promise<T> {
		const key = path.path();
		const cache = this.fileContentCache[key];
		const stats = await fs.promises.stat(path.path());

		if (cache && cache.mtime === stats.mtimeMs) {
			return cache.content;
		}

		const content = await cb();

		this.fileContentCache[key] = {
			content,
			exists: true,
			mtime: stats.mtimeMs,
		};

		return content;
	}

	async cacheFilesInDir(dir: Path, cb: () => Promise<Path[]>): Promise<Path[]> {
		const key = dir.path();

		if (this.dirFilesCache[key]) {
			return this.dirFilesCache[key];
		}

		const files = await cb();

		this.dirFilesCache[key] = files;

		return files;
	}

	clearFileCache() {
		this.fileContentCache = {};
	}

	clearFinderCache() {
		this.dirFilesCache = {};
	}

	getFileCache<T>(path: Path): FileCache<T> | null {
		return this.fileContentCache[path.path()] || null;
	}

	markMissingFile(path: Path): this {
		this.fileContentCache[path.path()] = {
			content: null,
			exists: false,
			mtime: 0,
		};

		return this;
	}
}
