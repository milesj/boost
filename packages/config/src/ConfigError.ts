import { createScopedError } from '@boost/internal';

const errors = {
	EXTENDS_ONLY_ROOT: 'Extends setting `{0}` must only be defined in a root config.',
	EXTENDS_UNKNOWN_PATH: 'Cannot extend configuration. Unknown module or file path "{0}".',
	LOADER_UNSUPPORTED: 'Unsupported loader format "{0}".',
	PACKAGE_UNKNOWN_SCOPE: 'Unable to determine package scope. No parent `package.json` found.',
	ROOT_INVALID: 'Invalid configuration root. Requires a `{0}` folder and `package.json`.',
	ROOT_NO_PACKAGE:
		'Config folder `{0}` found without a relative `package.json`. Both must be located in the project root.',
	ROOT_ONLY_OVERRIDES: 'Overrides setting `{0}` must only be defined in a root config.',
};

export type ConfigErrorCode = keyof typeof errors;

export const ConfigError = createScopedError<ConfigErrorCode>('CFG', 'ConfigError', errors);
