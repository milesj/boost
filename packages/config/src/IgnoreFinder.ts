import fs from 'fs';
import { Path } from '@boost/common';
import { Blueprint, Schemas } from '@boost/common/optimal';
import { color } from '@boost/internal';
import { Finder } from './Finder';
import { IgnoreFile, IgnoreFinderOptions } from './types';

export class IgnoreFinder extends Finder<IgnoreFile, IgnoreFinderOptions> {
	blueprint(schemas: Schemas): Blueprint<IgnoreFinderOptions> {
		const { string } = schemas;

		return {
			name: string().required().camelCase(),
		};
	}

	/**
	 * Find a single ignore file in the provided directory.
	 */
	async findFilesInDir(dir: Path): Promise<Path[]> {
		const files: Path[] = [];
		const path = dir.append(this.getFileName());

		if (path.exists()) {
			files.push(path);
		}

		this.debug.invariant(
			files.length > 0,
			`Finding ignore files in ${color.filePath(dir.path())}`,
			files.map((file) => file.name()).join(', '),
			'No files',
		);

		return Promise.resolve(files);
	}

	/**
	 * Return an ignore specific file.
	 */
	getFileName(): string {
		return `.${this.options.name.toLowerCase()}ignore`;
	}

	/**
	 * Load and parse a list of found files into a list of ignore patterns.
	 */
	async resolveFiles(basePath: Path, foundFiles: Path[]): Promise<IgnoreFile[]> {
		this.debug('Resolving %d ignore files', foundFiles.length);

		return Promise.all(
			foundFiles.map(async (filePath) => {
				const contents = await this.cache.cacheFileContents(filePath, () =>
					fs.promises.readFile(filePath.path(), 'utf8'),
				);
				const ignore = contents
					.split('\n')
					.map((line) => line.trim())
					.filter((line) => line !== '' && !line.startsWith('#'));

				return {
					ignore,
					path: filePath,
					source: this.isRootDir(filePath.parent(), true) ? ('root' as const) : ('branch' as const),
				};
			}),
		);
	}
}
