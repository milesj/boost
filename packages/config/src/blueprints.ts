import { predicates } from '@boost/common';
import { ExtendsSetting, PluginsSettingOption } from './types';

const { array, bool, object, string, union } = predicates;

export const extendsSetting = union<ExtendsSetting>(
  [string().notEmpty(), array(string().notEmpty())],
  [],
);

export const pluginsSetting = object(union<PluginsSettingOption>([bool(), object()], {}));
