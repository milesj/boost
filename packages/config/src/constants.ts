import { ExtType } from './types';

export const CONFIG_FOLDER = '.config';

export const PACKAGE_FILE = 'package.json';

export const DEFAULT_EXTS: ExtType[] = ['js', 'json', 'cjs', 'mjs', 'ts', 'json5', 'yaml', 'yml'];
