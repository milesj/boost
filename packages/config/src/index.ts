/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

export * from './ConfigError';
export * from './Configuration';
export * from './helpers/getEnv';
export * from './helpers/mergeArray';
export * from './helpers/mergeExtends';
export * from './helpers/mergeObject';
export * from './helpers/mergePlugins';
export * from './helpers/overwrite';
export * from './schemas';
export * from './types';
