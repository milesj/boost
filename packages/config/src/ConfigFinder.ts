/* eslint-disable no-param-reassign, no-await-in-loop */

import {
  Predicates,
  Path,
  PackageStructure,
  toArray,
  isModuleName,
  isFilePath,
} from '@boost/common';
import { RuntimeError } from '@boost/internal';
import minimatch from 'minimatch';
import loadCjs from './loaders/cjs';
import loadJs from './loaders/js';
import loadJson from './loaders/json';
import loadMjs from './loaders/mjs';
import loadYaml from './loaders/yaml';
import Finder from './Finder';
import createFileName from './helpers/createFileName';
import getEnv from './helpers/getEnv';
import { CONFIG_FOLDER, DEFAULT_EXTS, PACKAGE_FILE } from './constants';
import {
  ConfigFile,
  ConfigFinderOptions,
  ExtendsSetting,
  ExtType,
  OverridesSetting,
} from './types';

export default class ConfigFinder<T extends object> extends Finder<
  ConfigFile<T>,
  ConfigFinderOptions<T>
> {
  blueprint({ array, bool, func, shape, string }: Predicates) {
    return {
      extendsSetting: string(),
      extensions: array(string<ExtType>(), DEFAULT_EXTS),
      includeEnv: bool(true),
      loaders: shape({
        cjs: func(loadCjs).notNullable(),
        js: func(loadJs).notNullable(),
        json: func(loadJson).notNullable(),
        mjs: func(loadMjs).notNullable(),
        yaml: func(loadYaml).notNullable(),
        yml: func(loadYaml).notNullable(),
      }).exact(),
      name: string()
        .required()
        .camelCase(),
      overridesSetting: string(),
    };
  }

  /**
   * Determine a files package scope by finding the first parent `package.json`
   * when traversing up directories. We will leverage the cache as much as
   * possible for performance.
   *
   * @see https://nodejs.org/api/esm.html#esm_package_scope_and_file_extensions
   */
  async determinePackageScope(dir: Path): Promise<PackageStructure> {
    let currentDir = dir.isDirectory() ? dir : dir.parent();

    while (!this.isFileSystemRoot(currentDir)) {
      const pkgPath = currentDir.append(PACKAGE_FILE);
      const cache = this.cache.getFileCache<PackageStructure>(pkgPath);

      if (cache) {
        if (cache.exists) {
          return cache.content;
        }
        // Fall-through
      } else if (pkgPath.exists()) {
        return this.cache.cacheFileContents(pkgPath, () => loadJson(pkgPath));
      } else {
        this.cache.markMissingFile(pkgPath);
      }

      currentDir = currentDir.parent();
    }

    throw new RuntimeError('config', 'CFG_PACKAGE_SCOPE_UNKNOWN');
  }

  /**
   * Find all configuration and environment specific files in a directory
   * by looping through all the defined extension options.
   * Will only search until the first file is found, and will not return multiple extensions.
   */
  async findFilesInDir(dir: Path): Promise<Path[]> {
    const isRoot = this.isRootDir(dir);
    const baseDir = isRoot ? dir.append(CONFIG_FOLDER) : dir;

    return this.cache.cacheFilesInDir(baseDir, async () => {
      const paths: Path[] = [];

      // eslint-disable-next-line no-restricted-syntax
      for (const ext of this.options.extensions) {
        const files = [baseDir.append(this.getFileName(ext, !isRoot, false))];

        if (this.options.includeEnv) {
          files.push(baseDir.append(this.getFileName(ext, !isRoot, true)));
        }

        await Promise.all(
          files.map(configPath => {
            if (configPath.exists()) {
              paths.push(configPath);
            }

            return configPath;
          }),
        );

        // Once we find any file, we abort looking for others
        if (paths.length > 0) {
          break;
        }
      }

      return paths;
    });
  }

  /**
   * Create and return a config file name, with optional branch and environment variants.
   */
  getFileName(ext: string, isBranch: boolean, isEnv: boolean): string {
    const { name } = this.options;

    return createFileName(name, ext, {
      envSuffix: isEnv ? getEnv(name) : '',
      leadingDot: isBranch,
    });
  }

  /**
   * Load file and package contents from a list of file paths.
   * Extract and apply extended and override configs based on the base path.
   */
  async resolveFiles(basePath: Path, foundFiles: Path[]): Promise<ConfigFile<T>[]> {
    const configs = await Promise.all(foundFiles.map(filePath => this.loadConfig(filePath)));

    // Overrides take the highest precedence and must appear after everything,
    // including branch level configs. However, they must extract first so that
    // extends functionality can be inherited.
    if (this.options.overridesSetting) {
      const overriddenConfigs = await this.extractOverriddenConfigs(basePath, configs);

      if (overriddenConfigs.length > 0) {
        configs.push(...overriddenConfigs);
      }
    }

    // Configs that have been extended from root configs must
    // appear before everything else, in the order they were defined
    if (this.options.extendsSetting) {
      const extendedConfigs = await this.extractExtendedConfigs(configs);

      if (extendedConfigs.length > 0) {
        configs.unshift(...extendedConfigs);
      }
    }

    return configs;
  }

  /**
   * Extract a list of config files to extend, in order, from the list of previously loaded
   * config files, which is typically from the root. The list to extract can be located within
   * a property that matches the `extendsSetting` option.
   */
  protected extractExtendedConfigs(configs: ConfigFile<T>[]): Promise<ConfigFile<T>[]> {
    const { name, extendsSetting } = this.options;
    const extendsPaths: Path[] = [];

    configs.forEach(({ config, path, source }) => {
      const key = extendsSetting as keyof T;
      const extendsFrom = config[key] as ExtendsSetting | undefined;

      if (source === 'root' || source === 'override') {
        delete config[key];
      } else if (extendsFrom) {
        throw new RuntimeError('config', 'CFG_EXTENDS_ROOT_ONLY', [key]);
      } else {
        return;
      }

      toArray(extendsFrom).forEach(extendsPath => {
        if (isModuleName(extendsPath)) {
          extendsPaths.push(
            new Path(extendsPath, createFileName(name, 'js', { envSuffix: 'preset' })),
          );
        } else if (isFilePath(extendsPath)) {
          extendsPaths.push(path.parent().append(extendsPath));
        } else {
          throw new RuntimeError('config', 'CFG_EXTENDS_UNKNOWN_PATH', [extendsPath]);
        }
      });
    });

    return Promise.all(extendsPaths.map(path => this.loadConfig(path)));
  }

  /**
   * Extract all root config overrides that match the current path used to load with.
   * Overrides are located within a property that matches the `overridesSetting` option.
   */
  protected extractOverriddenConfigs(basePath: Path, configs: ConfigFile<T>[]): ConfigFile<T>[] {
    const { overridesSetting } = this.options;
    const overriddenConfigs: ConfigFile<T>[] = [];

    configs.forEach(({ config, path, source }) => {
      const key = overridesSetting as keyof T;
      const overrides = config[key] as OverridesSetting<T>[] | undefined;

      if (source === 'root') {
        delete config[key];
      } else if (overrides) {
        throw new RuntimeError('config', 'CFG_OVERRIDES_ROOT_ONLY', [key]);
      } else {
        return;
      }

      toArray(overrides).forEach(({ exclude, include, settings }) => {
        const excludePatterns = toArray(exclude);
        const includePatterns = toArray(include);
        const options = { dot: true, matchBase: true };

        if (
          includePatterns.some(pattern => minimatch(basePath.path(), pattern, options)) &&
          !excludePatterns.some(pattern => minimatch(basePath.path(), pattern, options))
        ) {
          overriddenConfigs.push({
            config: settings,
            path,
            source: 'override',
          });
        }
      });
    });

    return overriddenConfigs;
  }

  /**
   * Load config contents from the provided file path using one of the defined loaders.
   */
  protected async loadConfig(path: Path): Promise<ConfigFile<T>> {
    const pkg = await this.determinePackageScope(path);
    const config = await this.cache.cacheFileContents(path, async () => {
      const { loaders } = this.options;
      const ext = path.ext(true);

      switch (ext) {
        case 'cjs':
          return loaders.cjs(path, pkg);
        case 'js':
          return loaders.js(path, pkg);
        case 'json':
          return loaders.json(path, pkg);
        case 'mjs':
          // Not easily testable yet
          // istanbul ignore next
          return loaders.mjs(path, pkg);
        case 'yaml':
        case 'yml':
          return loaders.yaml(path, pkg);
        default:
          throw new RuntimeError('config', 'CFG_LOADER_UNSUPPORTED', [ext]);
      }
    });

    return {
      config,
      path,
      source: path.path().includes(CONFIG_FOLDER) ? 'root' : 'branch',
    };
  }
}
