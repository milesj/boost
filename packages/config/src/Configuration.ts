import { Contract, predicates, PortablePath, Blueprint } from '@boost/common';
import { Event, WaterfallEvent } from '@boost/event';
import Cache from './Cache';
import ConfigFinder from './ConfigFinder';
import IgnoreFinder from './IgnoreFinder';
import Processor from './Processor';
import {
  ProcessedConfig,
  ConfigFile,
  Handler,
  IgnoreFile,
  ConfigFinderOptions,
  ProcessorOptions,
} from './types';

export default abstract class Configuration<T extends object> extends Contract<T> {
  readonly onLoadedConfig = new WaterfallEvent<ConfigFile<T>[]>('loaded-config');

  readonly onLoadedIgnore = new WaterfallEvent<IgnoreFile[]>('loaded-ignore');

  readonly onProcessedConfig = new Event<[Required<T>]>('processed-config');

  private cache: Cache;

  private configFinder: ConfigFinder<T>;

  private ignoreFinder: IgnoreFinder;

  private processor: Processor<T>;

  constructor(name: string) {
    super();

    this.cache = new Cache();
    this.configFinder = new ConfigFinder({ name }, this.cache);
    this.ignoreFinder = new IgnoreFinder({ name }, this.cache);
    this.processor = new Processor({ name });
    this.bootstrap();
  }

  /**
   * Clear all cache.
   */
  clearCache(): this {
    this.clearFileCache();
    this.clearFinderCache();

    return this;
  }

  /**
   * Clear all cached file contents.
   */
  clearFileCache(): this {
    this.cache.clearFileCache();

    return this;
  }

  /**
   * Clear all cached directory and file path information.
   */
  clearFinderCache(): this {
    this.cache.clearFinderCache();

    return this;
  }

  /**
   * Traverse upwards from the branch directory, until the root directory is found,
   * or we reach to top of the file system. While traversing, find all config files
   * within each branch directory, and the root.
   */
  async loadConfigFromBranchToRoot(dir: PortablePath): Promise<ProcessedConfig<T>> {
    const configs = await this.getConfigFinder().loadFromBranchToRoot(dir);

    return this.processConfigs(this.onLoadedConfig.emit(configs));
  }

  /**
   * Load config files from the defined root. Root is determined by a relative
   * `.config` folder and `package.json` file.
   */
  async loadConfigFromRoot(dir: PortablePath = process.cwd()): Promise<ProcessedConfig<T>> {
    const configs = await this.getConfigFinder().loadFromRoot(dir);

    return this.processConfigs(this.onLoadedConfig.emit(configs));
  }

  /**
   * Traverse upwards from the branch directory, until the root directory is found,
   * or we reach to top of the file system. While traversing, find all ignore files
   * within each branch directory, and the root.
   */
  async loadIgnoreFromBranchToRoot(dir: PortablePath): Promise<IgnoreFile[]> {
    const ignores = await this.getIgnoreFinder().loadFromBranchToRoot(dir);

    return this.onLoadedIgnore.emit(ignores);
  }

  /**
   * Load ignore file from the defined root. Root is determined by a relative
   * `.config` folder and `package.json` file.
   */
  async loadIgnoreFromRoot(dir: PortablePath = process.cwd()): Promise<IgnoreFile[]> {
    const ignores = await this.getIgnoreFinder().loadFromRoot(dir);

    return this.onLoadedIgnore.emit(ignores);
  }

  /**
   * Add a process handler to customize the processing of key-value setting pairs.
   * May only run a processor on settings found in the root of the configuration object.
   */
  protected addProcessHandler<K extends keyof T, V = T[K]>(key: K, handler: Handler<V>): this {
    this.getProcessor().addHandler(key, handler);

    return this;
  }

  /**
   * Called on initialization.
   */
  protected bootstrap() {}

  /**
   * Configure the finder instance.
   */
  protected configureFinder(options: Omit<ConfigFinderOptions<T>, 'name'>): this {
    this.getConfigFinder().configure(options);

    return this;
  }

  /**
   * Configure the processor instance.
   */
  protected configureProcessor(options: Omit<ProcessorOptions, 'name'>): this {
    this.getProcessor().configure(options);

    return this;
  }

  /**
   * Return the config file finder instance.
   */
  protected getConfigFinder(): ConfigFinder<T> {
    return this.configFinder;
  }

  /**
   * Return the ignore file finder instance.
   */
  protected getIgnoreFinder(): IgnoreFinder {
    return this.ignoreFinder;
  }

  /**
   * Return the processor instance.
   */
  protected getProcessor(): Processor<T> {
    return this.processor;
  }

  /**
   * Process all loaded config objects into a single config object, and then validate.
   */
  protected async processConfigs(files: ConfigFile<T>[]): Promise<ProcessedConfig<T>> {
    const config = await this.getProcessor().process(
      this.options,
      files,
      this.blueprint(predicates) as Blueprint<T>,
    );

    this.onProcessedConfig.emit([config]);

    return {
      config,
      files,
    };
  }
}
