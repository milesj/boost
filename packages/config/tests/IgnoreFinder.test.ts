import { EOL } from 'os';
import { normalizeSeparators } from '@boost/common/test';
import { getFixturePath } from '@boost/test-utils';
import { Cache } from '../src/Cache';
import { IgnoreFinder } from '../src/IgnoreFinder';
import { mockSystemPath } from './helpers';

describe('IgnoreFinder', () => {
	let cache: Cache;
	let finder: IgnoreFinder;

	beforeEach(() => {
		cache = new Cache();
		finder = new IgnoreFinder({ name: 'boost' }, cache);
	});

	it('errors if name is not in camel case', () => {
		expect(() => {
			finder = new IgnoreFinder({ name: 'Boosty Boost' }, cache);
		}).toThrowErrorMatchingSnapshot();
	});

	it('caches root, file, and directory information', async () => {
		const tempRoot = getFixturePath('config-ignore-file-tree');

		await finder.loadFromBranchToRoot(
			normalizeSeparators(`${tempRoot}/src/app/feature/signup/flow`),
		);

		expect(cache.rootDir).toEqual(mockSystemPath(tempRoot));
		expect(cache.configDir).toEqual(mockSystemPath(`${tempRoot}/.config`));
		expect(cache.pkgPath).toEqual(mockSystemPath(`${tempRoot}/package.json`));
		expect(cache.fileContentCache).toEqual({
			[mockSystemPath(`${tempRoot}/.boostignore`).path()]: {
				content: `*.log${EOL}*.lock`,
				exists: true,
				mtime: expect.any(Number),
			},
			[mockSystemPath(`${tempRoot}/src/app/feature/.boostignore`).path()]: {
				content: `# Compiled${EOL}lib/`,
				exists: true,
				mtime: expect.any(Number),
			},
			[mockSystemPath(`${tempRoot}/src/app/feature/signup/.boostignore`).path()]: {
				content: '# Empty',
				exists: true,
				mtime: expect.any(Number),
			},
		});
	});

	describe('loadFromBranchToRoot()', () => {
		it('returns all ignore files from a target file', async () => {
			const tempRoot = getFixturePath('config-ignore-file-tree');

			const files = await finder.loadFromBranchToRoot(
				normalizeSeparators(`${tempRoot}/src/app/components/build/Button.tsx`),
			);

			expect(files).toEqual([
				{
					ignore: ['*.log', '*.lock'],
					path: mockSystemPath(`${tempRoot}/.boostignore`),
					source: 'root',
				},
				{
					ignore: ['esm/'],
					path: mockSystemPath(`${tempRoot}/src/app/components/build/.boostignore`),
					source: 'branch',
				},
			]);
		});

		it('returns all ignore files from a target folder', async () => {
			const tempRoot = getFixturePath('config-ignore-file-tree');

			const files = await finder.loadFromBranchToRoot(
				normalizeSeparators(`${tempRoot}/src/app/feature/signup/flow/`),
			);

			expect(files).toEqual([
				{
					ignore: ['*.log', '*.lock'],
					path: mockSystemPath(`${tempRoot}/.boostignore`),
					source: 'root',
				},
				{
					ignore: ['lib/'],
					path: mockSystemPath(`${tempRoot}/src/app/feature/.boostignore`),
					source: 'branch',
				},
				{
					ignore: [],
					path: mockSystemPath(`${tempRoot}/src/app/feature/signup/.boostignore`),
					source: 'branch',
				},
			]);
		});
	});

	describe('loadFromRoot()', () => {
		it('returns ignore file from root folder', async () => {
			const tempRoot = getFixturePath('config-ignore-file-tree');

			const files = await finder.loadFromRoot(tempRoot);

			expect(files).toEqual([
				{
					ignore: ['*.log', '*.lock'],
					path: mockSystemPath(`${tempRoot}/.boostignore`),
					source: 'root',
				},
			]);
		});

		it('errors if root folder is missing a `package.json`', async () => {
			const tempRoot = getFixturePath('config-root-without-package-json');

			await expect(finder.loadFromRoot(tempRoot)).rejects.toThrow(
				'Config folder `.config` found without a relative `package.json`. Both must be located in the project root.',
			);
		});
	});
});
