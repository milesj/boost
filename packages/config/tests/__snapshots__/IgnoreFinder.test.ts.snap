// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IgnoreFinder errors if name is not in camel case 1`] = `
"The following validations have failed for \`IgnoreFinder\`:
  - Invalid field \\"name\\". String must be in camel case. (pattern \\"^[a-z][a-zA-Z0-9]+$\\")"
`;
